src/Akka.Streams.Kafka/Stages/Producers/TransactionalProducerStage.cs

```diff
@@ -0,0 +1,269 @@
+        }
+
+        private void ResumeDemand(bool tryToPull = true)
+        {
+            SetHandler(_stage.Out, onPull: () => TryPull(_stage.In));
+
+            _demandSuspended = false;
+
+            if (tryToPull && IsAvailable(_stage.Out) && !HasBeenPulled(_stage.In))
+            {
+                TryPull(_stage.In);
+            }
+        }
+
+        private void SuspendDemand()
+        {
+            if (!_demandSuspended)
+            {
+                // suspend demand while a commit is in process so we can drain any outstanding message acknowledgements
+                SetHandler(_stage.Out, onPull: () => { });
+            }
+
+            _demandSuspended = true;
+        }
+
+        protected override void OnTimer(object timerKey)
+        {
+            if (timerKey.Equals(CommitSchedulerKey))
+                MaybeCommitTransaction();
+        }
+
+        private void MaybeCommitTransaction(bool beginNewTransaction = true)
+        {
+            var awaitingConf = AwaitingConfirmation.Current;
+            if (_batchOffsets is NonemptyTransactionBatch nonemptyTransactionBatch && awaitingConf == 0)
+            {
+                CommitTransaction(nonemptyTransactionBatch, beginNewTransaction);
+            }
+            else if (awaitingConf > 0)
+            {
+                SuspendDemand();
+                ScheduleOnce(CommitSchedulerKey, _messageDrainInterval);
+            }
+            else
+            {
+                ScheduleOnce(CommitSchedulerKey, _commitInterval);
+            }
+        }
+
+        protected override void PostSend(IEnvelope<K, V, TPassThrough> msg)
+        {
+            var marker = msg.PassThrough as PartitionOffsetCommittedMarker;
+            if (marker != null)
+                _batchOffsets = _batchOffsets.Updated(marker);
+        }
+
+        public override void OnCompletionSuccess()
+        {
+            Log.Debug("Comitting final transaction before shutdown");
+            CancelTimer(CommitSchedulerKey);
+            MaybeCommitTransaction(beginNewTransaction: false);
+            base.OnCompletionSuccess();
+        }
+
+        public override void OnCompletionFailure(Exception ex)
+        {
+            Log.Debug("Aborting transaction due to stage failure");
+            AbortTransaction();
+            _batchOffsets.CommittingFailed();
+            base.OnCompletionFailure(ex);
+        }
+
+        private void CommitTransaction(NonemptyTransactionBatch batch, bool beginNewTransaction)
+        {
+            var groupId = batch.GroupId;
+            Log.Debug("Committing transaction for consumer group '{0}' with offsets: {1}", groupId, batch.Offsets);
+            var offsetMap = batch.OffsetMap();
+            
+            // TODO: Add producer work with transactions
+            /* scala code:
+             producer.sendOffsetsToTransaction(offsetMap.asJava, group)
+             producer.commitTransaction()
+             */
```

Comment:
Here is an example of missing lines, that should be replaced to close #85 . Few more places like this in next methods.